miner/worker.go

```diff
@@ -1201,64 +1205,177 @@ func (w *worker) commit(uncles []*types.Header, interval func(), update bool, st
 	return nil
 }
 
-func (w *worker) findMostProfitableBundle(bundles []types.Transactions, coinbase common.Address, parent *types.Block, header *types.Header) (types.Transactions, *big.Int, *big.Int, uint64) {
-	maxBundlePrice := new(big.Int)
-	maxTotalEth := new(big.Int)
-	var maxTotalGasUsed uint64
-	maxBundle := types.Transactions{}
+type simulatedBundle struct {
+	mevGasPrice       *big.Int
+	totalEth          *big.Int
+	ethSentToCoinbase *big.Int
+	totalGasUsed      uint64
+	originalBundle    types.MevBundle
+}
+
+func (w *worker) generateFlashbotsBundle(bundles []types.MevBundle, coinbase common.Address, parent *types.Block, header *types.Header, pendingTxs map[common.Address]types.Transactions) (types.Transactions, simulatedBundle, int, error) {
+	simulatedBundles, err := w.simulateBundles(bundles, coinbase, parent, header, pendingTxs)
+	if err != nil {
+		return nil, simulatedBundle{}, 0, err
+	}
+
+	sort.SliceStable(simulatedBundles, func(i, j int) bool {
+		return simulatedBundles[j].mevGasPrice.Cmp(simulatedBundles[i].mevGasPrice) < 0
+	})
+
+	return w.mergeBundles(simulatedBundles, parent, header, pendingTxs)
+}
+
+func (w *worker) mergeBundles(bundles []simulatedBundle, parent *types.Block, header *types.Header, pendingTxs map[common.Address]types.Transactions) (types.Transactions, simulatedBundle, int, error) {
+	finalBundle := types.Transactions{}
+
+	state, err := w.chain.StateAt(parent.Root())
+	if err != nil {
+		return nil, simulatedBundle{}, 0, err
+	}
+	gasPool := new(core.GasPool).AddGas(header.GasLimit)
+
+	prevState := state
+	prevGasPool := gasPool
+
+	mergedBundle := simulatedBundle{
+		totalEth:          new(big.Int),
+		ethSentToCoinbase: new(big.Int),
+	}
+
+	count := 0
 	for _, bundle := range bundles {
-		if len(bundle) == 0 {
+		prevState = state.Copy()
+		prevGasPool = new(core.GasPool).AddGas(gasPool.Gas())
+
+		simmed, err := w.computeBundleGas(bundle.originalBundle, parent, header, state, gasPool, pendingTxs, len(finalBundle))
+		if err != nil || simmed.totalEth.Cmp(new(big.Int)) < 0 {
+			state = prevState
+			gasPool = prevGasPool
 			continue
 		}
-		totalEth, totalGasUsed, err := w.computeBundleGas(bundle, parent, header)
 
+		log.Info("Included bundle", "ethToCoinbase", ethIntToFloat(simmed.totalEth), "gasUsed", simmed.totalGasUsed, "bundleScore", simmed.mevGasPrice, "bundleLength", len(simmed.originalBundle.Txs), "worker", w.flashbots.maxMergedBundles)
+
+		finalBundle = append(finalBundle, bundle.originalBundle.Txs...)
+		mergedBundle.totalEth.Add(mergedBundle.totalEth, simmed.totalEth)
+		mergedBundle.ethSentToCoinbase.Add(mergedBundle.ethSentToCoinbase, simmed.ethSentToCoinbase)
+		mergedBundle.totalGasUsed += simmed.totalGasUsed
+		count++
+
+		if count >= w.flashbots.maxMergedBundles {
+			break
+		}
+	}
+
+	if len(finalBundle) == 0 || count != w.flashbots.maxMergedBundles {
```

Comment:
Why checking `count != w.flashbots.maxMergedBundles` here? Won't it be allowed to have fewer bundles than the set max?

Reply:
There is a many workers running in parallel, each with `w.flashbots.maxMergedBundles` set to `0...miner.maxmergedbundles`. This line just prevents these workers from doing duplicate work, since there must exist a worker where `count == w.flashbots.maxMergedBundles`

Reply:
Oh right, got it. Thanks!